File: Longest Common Subsequence/Longest Common Subsequence.cpp
/// This code is contributed by @iuliagroza

/// Time Complexity: O(n*m), where n is the length of the A sequence and m the length of the B sequence.
/// Additional Space: O(n*m)

#include <iostream>
#include <vector>

#define nMax 1050

int lcs[nMax][nMax];  /// Declared globally just to not do stack overflow

int main()
{
    std::ios_base::sync_with_stdio(0);
    std::cin.tie(0);
    std::cout.tie(0);
    int n, m;
    std::cin >> n >> m;

    int A[nMax];
    for(int i=1; i<=n; i++)
        std::cin >> A[i];

    int B[nMax];
    for(int i=1; i<=m; i++)
        std::cin >> B[i];

    for(int i=1; i<=n; i++)  /// Applying the recurrence. The DP structure being declared globally, it is already initialized with 0
        for(int j=1; j<=m; j++)
            if(A[i]==B[j])
                lcs[i][j] = lcs[i-1][j-1]+1;
            else
                lcs[i][j] = std::max(lcs[i-1][j], lcs[i][j-1]);

    int i=n, j=m;
    std::vector<int> sol;
    while(i && j)  /// Building the solution vector
    {
        if(A[i]==B[j])
        {
            sol.push_back(A[i]);  /// It is inversely memorated because we have to start from the maximum LCS
            i--;
            j--;
        }
        else if(lcs[i-1][j] < lcs[i][j-1])
            j--;
        else
            i--;
    }

    std::cout << sol.size() << "\n";
    std::vector<int>::reverse_iterator it;
    for(it=sol.rbegin(); it!=sol.rend(); it++)  /// Inversely output the vector to show it in ascending order
        std::cout << *it << " ";
    return 0;
}

File: Longest Common Subsequence/Longest Common Subsequence.py
# This code is contributed by @iuliagroza

# Time Complexity: O(n*m), where n is the length of the A sequence and m the length of the B sequence.
# Additional Space: O(n*m)

n = 5
A = [1, 7, 3, 9, 8]

m = 3
B = [7, 5, 8]

lcs = []
for i in range(0, n+1):
    lcs.append([]);
    for j in range(0, m+1):
        lcs[i].append(0)

for i in range(1, n+1):  # Building the DP structure using the recurrence
    for j in range(1, m+1):
        if A[i-1]==B[j-1]:
            lcs[i][j] = lcs[i-1][j-1]+1;
        else:
            lcs[i][j] = max(lcs[i-1][j], lcs[i][j-1])

i=n
j=m
sol = []
while i>0 and j>0:  # Build the solution array starting from the maximum (in reversed order)
    if A[i-1]==B[j-1]:
        sol.append(A[i-1])
        i -= 1
        j -= 1
    elif lcs[i-1][j] < lcs[i][j-1]:
        j -= 1
    else:
        i -= 1

print(len(sol))
sol = sol[::-1]  # Reverse the solution array to print it in ascending order
print(sol)